File: stream-processing/src/main/scala/Entities.scala
package org.tcc2.streaming

import play.api.libs.json.JsObject

trait Event {
  def eventId: String
  def timestamp: String
  def userId: Int
  def eventType: String
}

case class AdImpressionEvent(
  eventId: String,
  timestamp: String,
  userId: Int,
  campaignId: String,
  adPlatform: String,
  device: String,
  eventType: String = "ad_impression"
) extends Event

case class AdClickEvent(
  eventId: String,
  timestamp: String,
  userId: Int,
  campaignId: String,
  adPlatform: String,
  landingPage: String,
  eventType: String = "ad_click"
 ) extends Event

case class LandingPageViewEvent(
  eventId: String,
  timestamp: String,
  userId: Int,
  campaignId: String,
  landingPage: String,
  referrer: String,
  eventType: String = "landing_page_view"
) extends Event

case class AddToCartEvent(
  eventId: String,
  timestamp: String,
  userId: Int,
  productId: String,
  eventType: String = "add_to_cart"
) extends Event

case class CheckoutStartEvent(
  eventId: String,
  timestamp: String,
  userId: Int,
  cartItems: Array[String],
  eventType: String = "checkout_start",
  email: Option[String]
) extends Event

case class PurchaseEvent(
  eventId: String,
  timestamp: String,
  userId: Int,
  orderId: String,
  eventType: String = "purchase"
) extends Event

case class FlattenedEvent(
   event_id: String,
   timestamp: String,
   user_id: Int,
   event_type: String,
   event_data: JsObject
 )

File: datasources/entities.py
from typing import TypedDict, List, Optional

class AdImpressionEvent(TypedDict):
    event_id: str
    timestamp: str
    event_type: str = "ad_impression"
    campaign_id: str
    ad_platform: str
    device: str

class AdClickEvent(TypedDict):
    event_id: str
    timestamp: str
    event_type: str = "ad_click"
    campaign_id: str
    ad_platform: str
    landing_page: str

class LandingPageViewEvent(TypedDict):
    event_id: str
    timestamp: str
    event_type: str = "landing_page_view"
    campaign_id: str
    landing_page: str
    referrer: str

class AddToCartEvent(TypedDict):
    event_id: str
    timestamp: str
    event_type: str = "add_to_cart"
    product_id: str

class CheckoutStartEvent(TypedDict):
    event_id: str
    timestamp: str
    event_type: str = "checkout_start"
    cart_items: List[str]
    email: Optional[str]

class PurchaseEvent(TypedDict):
    event_id: str
    timestamp: str
    event_type: str = "purchase"
    order_id: str
    payment_method: str
